File: playground/web/verify.js
// Catalyst — runtime verification of agent-emitted A2UI surfaces against a catalog.
// Mirror of src/verify.py. Enforces output/catalog.json on a surface: every node is
// validated against the catalog's anyComponent / component schemas (unevaluatedProperties:false
// + const/enum discriminated union catches unknown components, bad variants, missing required
// props, and extra/reworded props).
//
// Validation engine: AJV (draft 2020-12) when a global is available (window.Ajv2020 /
// window.ajv2020 / Ajv) — load it e.g. from a CDN before this file. Without AJV it falls back
// to a focused validator that performs the same const/enum/required/closed-property checks.
//
// Failure policy is split by consequence (x-consequential on the catalog component, mirrored
// in output/consequential.json):
//   * non-consequential fails -> REPAIR (drop unknown props / snap bad variant to default enum),
//                                re-validate, keep if valid else drop. Logged.
//   * consequential fails     -> REJECT. Never coerced. Replaced with an inline error node.
//
//   verifySurface(surface, catalog, { lenient, consequential }) ->
//     { ok, lenient, results:[{component,consequential,valid,action,errors}], rejected:[...],
//       repaired:<surface> }
(function (root, factory) {
  const api = factory();
  if (typeof module !== "undefined" && module.exports) module.exports = api;   // node / tests
  if (typeof window !== "undefined") window.verifySurface = api.verifySurface;  // browser global
})(this, function () {
  // A2UI common props allowed on every node (ComponentCommon + CatalogComponentCommon).
  // unevaluatedProperties:false makes each component's prop set closed except for these.
  const COMMON_PROPS = ["component", "id", "children", "child", "weight"];

  // Stub of the shared A2UI common_types the catalog $refs (only the shapes we need).
  const COMMON_TYPES = {
    $id: "https://a2ui.org/specification/v0_9/common_types.json",
    $defs: {
      ComponentCommon: {
        type: "object",
        properties: {
          component: { type: "string" },
          id: { type: "string" },
          children: { type: "array", items: { type: ["string", "object"] } },
          child: { type: ["string", "object"] },
        },
      },
      DynamicString: { type: ["string", "object"] },
    },
  };

  function innerSchema(comp) {
    const allOf = (comp && comp.allOf) || [];
    for (const sub of allOf) if (sub && sub.properties) return sub;
    return {};
  }
  function componentInfo(catalog, name) {
    const comp = catalog.components && catalog.components[name];
    if (!comp) return null;
    const inner = innerSchema(comp);
    const props = inner.properties || {};
    const enums = {};
    for (const p in props) if (props[p] && Array.isArray(props[p].enum)) enums[p] = props[p].enum;
    const allowed = new Set(Object.keys(props).concat(COMMON_PROPS));
    return { required: inner.required || [], enums, allowed };
  }
  function consequentialSet(catalog, override) {
    if (override) return new Set(override);
    const s = new Set();
    const comps = catalog.components || {};
    for (const n in comps) if (comps[n] && comps[n]["x-consequential"] === true) s.add(n);
    return s;
  }

  // ---- validation engine: AJV if available, else a focused mirror ----
  function getAjv() {
    const g = (typeof window !== "undefined" && window) || (typeof globalThis !== "undefined" && globalThis) || {};
    return g.Ajv2020 || (g.ajv2020 && (g.ajv2020.default || g.ajv2020)) || g.Ajv || null;
  }
  function makeValidator(catalog) {
    const Ajv = getAjv();
    if (Ajv) {
      const ajv = new Ajv({ strict: false, allErrors: true });
      ajv.addSchema(COMMON_TYPES);
      ajv.addSchema(catalog);
      const fn = ajv.compile({ $ref: catalog.$id + "#/$defs/anyComponent" });
      return (node) => (fn(node) ? [] : (fn.errors || []).map((e) => (e.instancePath || "") + " " + e.message));
    }
    // fallback: same checks the schema would enforce
    return (node) => {
      const name = node.component;
      const info = componentInfo(catalog, name);
      if (!info) return ["unknown component '" + name + "'"];
      const errs = [];
      for (const r of info.required) if (!(r in node)) errs.push("missing required property '" + r + "'");
      for (const p in info.enums) if (p in node && info.enums[p].indexOf(node[p]) === -1)
        errs.push("'" + p + "' must be one of " + JSON.stringify(info.enums[p]));
      for (const k in node) if (!info.allowed.has(k)) errs.push("unevaluated (unknown) property '" + k + "'");
      return errs;
    };
  }

  function errorNode(name, why) {
    return {
      component: "Alert", tone: "danger",
      title: "Blocked: " + (name || "unknown"),
      description: (name || "component") + " failed catalog verification and was not rendered (" + why + ").",
    };
  }
  function repair(node, info) {
    const log = [];
    for (const k of Object.keys(node)) if (!info.allowed.has(k)) { delete node[k]; log.push("dropped unknown prop '" + k + "'"); }
    for (const p in info.enums) {
      const allowed = info.enums[p];
      if (p in node && allowed.indexOf(node[p]) === -1 && allowed.length) {
        log.push("snapped '" + p + "' '" + node[p] + "' -> '" + allowed[0] + "'");
        node[p] = allowed[0];
      }
    }
    const missing = info.required.filter((r) => !(r in node));
    return { log, missing };
  }

  function verifySurface(surface, catalog, opts) {
    opts = opts || {};
    const lenient = !!opts.lenient;
    // passUnknown: components the catalog doesn't define are kept as-is (rendered by a superset
    // renderer) instead of dropped. Lets a small "consequential contract" catalog run against a
    // rich renderer — only the consequential components it DOES define are enforced.
    const passUnknown = !!opts.passUnknown;
    const validate = makeValidator(catalog);
    const conseq = consequentialSet(catalog, opts.consequential);
    const results = [], rejected = [];

    function process(value) {
      if (Array.isArray(value)) {
        const out = [];
        for (const item of value) { const r = process(item); if (r !== null && r !== undefined) out.push(r); }
        return out;
      }
      if (value && typeof value === "object" && value.component) {
        const node = JSON.parse(JSON.stringify(value));
        const name = node.component;
        const isConseq = conseq.has(name);
        const known = componentInfo(catalog, name) !== null;
        // A component the catalog doesn't define: keep it (rich renderer handles it) when passUnknown.
        if (!known && passUnknown) {
          results.push({ component: name, consequential: false, valid: true, action: "keep", errors: [] });
          for (const k of Object.keys(node)) {
            const v = node[k];
            if (Array.isArray(v) || (v && typeof v === "object")) node[k] = process(v);
          }
          return node;
        }
        let errors = validate(node);
        let action;
        if (errors.length === 0) action = "keep";
        else if (isConseq) action = "reject";
        else if (!known) action = "drop";
        else {
          const info = componentInfo(catalog, name);
          const rep = repair(node, info);
          const errors2 = validate(node);
          if (errors2.length === 0 && rep.missing.length === 0) {
            action = "repair";
            errors = errors.concat(rep.log.map((m) => ">> " + m));
          } else action = "drop";
        }
        results.push({ component: name, consequential: isConseq,
          valid: action === "keep", action, errors });

        if (action === "reject") { rejected.push({ component: name, errors }); return errorNode(name, errors.join("; ") || "invalid"); }
        if (action === "drop") return null;
        for (const k of Object.keys(node)) {
          const v = node[k];
          if (Array.isArray(v) || (v && typeof v === "object")) node[k] = process(v);
        }
        return node;
      }
      if (value && typeof value === "object") {
        const out = {};
        for (const k in value) out[k] = process(value[k]);
        return out;
      }
      return value;
    }

    const repaired = process(surface);
    const anyInvalid = results.some((r) => r.action !== "keep");
    const ok = lenient ? rejected.length === 0 : !anyInvalid;
    return { ok, lenient, results, rejected, repaired };
  }

  return { verifySurface, COMMON_TYPES, componentInfo, consequentialSet };
});

File: src/verify.py
#!/usr/bin/env python3
"""
Catalyst — runtime verification of agent-emitted A2UI surfaces against a catalog.

This is the Python mirror of playground/web/verify.js. Both enforce output/catalog.json
on a surface (a list of A2UI component nodes) using real JSON-Schema validation
(unevaluatedProperties:false + const/enum discriminated union catches unknown components,
bad variants, missing required props, and extra/reworded props).

Failure policy is split by consequence (a component is consequential when the catalog
carries `x-consequential: true`, mirrored in output/consequential.json):

  * non-consequential node fails  -> REPAIR (drop unknown props / snap a bad variant to its
                                     default enum value), re-validate, keep if it now passes,
                                     otherwise drop it. Always logged.
  * consequential node fails      -> REJECT. Never coerced. The node is not rendered; a clear
                                     error is surfaced in its place.

`verify(surface, catalog, lenient=...)` returns a structured report plus a `repaired`
surface (used by lenient/playground callers). Strict callers treat any failure as fatal.
"""
import copy
import json
from pathlib import Path

try:
    from jsonschema import Draft202012Validator
    from referencing import Registry, Resource
    from referencing.jsonschema import DRAFT202012
    _HAVE_JSONSCHEMA = True
except Exception:  # pragma: no cover - fallback path when jsonschema is absent
    _HAVE_JSONSCHEMA = False

# Common props every A2UI node may carry (A2UI ComponentCommon + CatalogComponentCommon).
# unevaluatedProperties:false makes the per-component prop set closed *except* for these.
COMMON_PROPS = {"component", "id", "children", "child", "weight"}

# Stub of the shared A2UI common_types the catalog $refs. The real spec lives at
# https://a2ui.org/... ; for validation we only need the shapes these refs constrain.
COMMON_TYPES = {
    "$id": "https://a2ui.org/specification/v0_9/common_types.json",
    "$defs": {
        "ComponentCommon": {
            "type": "object",
            "properties": {
                "component": {"type": "string"},
                "id": {"type": "string"},
                "children": {"type": "array", "items": {"type": ["string", "object"]}},
                "child": {"type": ["string", "object"]},
            },
        },
        "DynamicString": {"type": ["string", "object"]},
    },
}


# ---------------------------------------------------------------------------
# catalog introspection
# ---------------------------------------------------------------------------
def _inner_schema(comp):
    """The component-specific allOf branch (the one that declares `properties`)."""
    for sub in comp.get("allOf", []):
        if isinstance(sub, dict) and "properties" in sub:
            return sub
    return {}


def component_info(catalog, name):
    """Allowed props, required list, and enum maps for one catalog component."""
    comp = catalog.get("components", {}).get(name)
    if comp is None:
        return None
    inner = _inner_schema(comp)
    props = inner.get("properties", {})
    enums = {p: spec["enum"] for p, spec in props.items()
             if isinstance(spec, dict) and "enum" in spec}
    return {
        "required": [r for r in inner.get("required", [])],
        "enums": enums,
        "allowed": set(props.keys()) | COMMON_PROPS,
    }


def consequential_set(catalog, sidecar=None):
    """Components flagged consequential — from x-consequential, or an override list."""
    if sidecar is not None:
        return set(sidecar)
    return {n for n, c in catalog.get("components", {}).items()
            if isinstance(c, dict) and c.get("x-consequential") is True}


# ---------------------------------------------------------------------------
# validation engine
# ---------------------------------------------------------------------------
def _build_validator(catalog):
    if not _HAVE_JSONSCHEMA:
        return None
    res_cat = Resource.from_contents(catalog, default_specification=DRAFT202012)
    res_com = Resource.from_contents(COMMON_TYPES, default_specification=DRAFT202012)
    registry = (Registry()
                .with_resource(catalog["$id"], res_cat)
                .with_resource(COMMON_TYPES["$id"], res_com))
    return Draft202012Validator(
        {"$ref": catalog["$id"] + "#/$defs/anyComponent"}, registry=registry)


def _node_errors(validator, catalog, node):
    """Return a list of human-readable validation errors for a single node ([] = valid)."""
    if validator is not None:
        return [e.message for e in validator.iter_errors(node)]
    # dependency-free fallback: mirror the schema checks the AJV/jsonschema path performs.
    name = node.get("component")
    info = component_info(catalog, name)
    if info is None:
        return ["unknown component '%s'" % name]
    errs = []
    for r in info["required"]:
        if r not in node:
            errs.append("missing required property '%s'" % r)
    for prop, allowed in info["enums"].items():
        if prop in node and node[prop] not in allowed:
            errs.append("'%s' must be one of %s" % (prop, allowed))
    for k in node:
        if k not in info["allowed"]:
            errs.append("unevaluated (unknown) property '%s'" % k)
    return errs


def _error_node(name, why):
    return {
        "component": "Alert", "tone": "danger",
        "title": "Blocked: %s" % (name or "unknown"),
        "description": "%s failed catalog verification and was not rendered (%s)."
                       % (name or "component", why),
    }


def _repair(node, info):
    """Non-consequential repair: drop unknown props, snap bad enums. Returns (log, missing)."""
    log = []
    for k in list(node.keys()):
        if k not in info["allowed"]:
            del node[k]
            log.append("dropped unknown prop '%s'" % k)
    for prop, allowed in info["enums"].items():
        if prop in node and node[prop] not in allowed and allowed:
            log.append("snapped '%s' %r -> '%s'" % (prop, node[prop], allowed[0]))
            node[prop] = allowed[0]
    missing = [r for r in info["required"] if r not in node]
    return log, missing


# ---------------------------------------------------------------------------
# public API
# ---------------------------------------------------------------------------
def verify(surface, catalog, lenient=False, consequential=None):
    """Verify a surface (node or list of nodes) against a catalog.

    Returns {ok, lenient, results:[...], rejected:[...], repaired:<surface>}.
    In lenient mode `repaired` is the surface to render (repairs applied, unknown/unrepairable
    nodes dropped, consequential failures replaced with an inline error node).
    """
    validator = _build_validator(catalog)
    conseq = consequential_set(catalog, consequential)
    results, rejected = [], []

    def process(value):
        if isinstance(value, list):
            out = []
            for item in value:
                r = process(item)
                if r is not None:
                    out.append(r)
            return out
        if isinstance(value, dict) and value.get("component"):
            node = copy.deepcopy(value)
            name = node.get("component")
            is_conseq = name in conseq
            errors = _node_errors(validator, catalog, node)
            known = component_info(catalog, name) is not None

            if not errors:
                action = "keep"
            elif is_conseq:
                action = "reject"
            elif not known:
                action = "drop"           # unknown component: nothing to repair
            else:
                info = component_info(catalog, name)
                log, missing = _repair(node, info)
                errors2 = _node_errors(validator, catalog, node)
                if not errors2 and not missing:
                    action = "repair"
                    errors = errors + [">> " + m for m in log]
                else:
                    action = "drop"

            results.append({"component": name, "consequential": is_conseq,
                            "valid": not errors if action == "keep" else False,
                            "action": action, "errors": errors})

            if action == "reject":
                rejected.append({"component": name, "errors": errors})
                return _error_node(name, "; ".join(errors) or "invalid")
            if action == "drop":
                return None
            # keep / repair: also process any nested nodes it carries
            for k, v in list(node.items()):
                if isinstance(v, (list, dict)):
                    node[k] = process(v)
            return node
        if isinstance(value, dict):
            return {k: process(v) for k, v in value.items()}
        return value

    repaired = process(surface)
    any_invalid = any(r["action"] != "keep" for r in results)
    ok = (len(rejected) == 0) if lenient else (not any_invalid)
    return {"ok": ok, "lenient": lenient, "results": results,
            "rejected": rejected, "repaired": repaired}


def load_catalog(path="output/catalog.json"):
    return json.loads(Path(path).read_text())


if __name__ == "__main__":
    import sys
    cat = load_catalog(sys.argv[1] if len(sys.argv) > 1 else "output/catalog.json")
    data = json.loads(sys.stdin.read())
    print(json.dumps(verify(data, cat, lenient=("--lenient" in sys.argv)), indent=2))
